File: leetcode/01-Arrays_and_Strings/0053-MaximumSubarray/MaximumSubarray1.py
# official solution
# 贪心算法：找出各个独立区域的最大和，比较这些和
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        curr_sum = max_sum = -20000000000
        for  i in range( len( nums ) ) :
            curr_sum = max( nums[i], nums[i]+curr_sum );
            max_sum = max( max_sum, curr_sum );
        return max_sum



        

File: leetcode/01-Arrays_and_Strings/0053-MaximumSubarray/MaximumSubarray1.cpp
#include <vector>
#include <iostream>

using namespace std;

// no need to find out the exact subarray
// greedy algorithm

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_sum = nums[0];
        int prev_sum = nums[0];
        int curr_sum = nums[0];
        
        for ( int i = 1; i < nums.size(); i++ ){
            
            if ( prev_sum < 0){
                curr_sum = nums[i];
                prev_sum = nums[i];
            } else {
                curr_sum = prev_sum + nums[i];
                prev_sum += nums[i];
            }

            if ( curr_sum > max_sum ) {
                max_sum = curr_sum;
            }
        }

        return max_sum;
    }
};

int main( int argc, char** argv ) {
    vector<int> nums = { -2, 1, -3, 4, -1, 2, 1, -5, 4};
    Solution s;
    cout << s.maxSubArray( nums );
    
}